server/routes/serviceProviderReferrals/serviceProviderReferralsController.ts

```diff
@@ -434,15 +435,49 @@ export default class ServiceProviderReferralsController {
     return ControllerUtils.renderWithLayout(res, view, serviceUser)
   }
 
-  async editSession(req: Request, res: Response): Promise<void> {
+  async editActionPlanSession(req: Request, res: Response): Promise<void> {
     const sessionNumber = Number(req.params.sessionNumber)
+    const actionPlan = await this.interventionsService.getActionPlan(res.locals.user.token.accessToken, req.params.id)
+
+    await this.scheduleAppointment(req, res, {
+      getReferral: () =>
+        this.interventionsService.getSentReferral(res.locals.user.token.accessToken, actionPlan.referralId),
+      getCurrentAppointment: () =>
+        this.interventionsService.getActionPlanAppointment(
+          res.locals.user.token.accessToken,
+          req.params.id,
+          sessionNumber
+        ),
+      scheduleAppointment: paramsForUpdate =>
+        this.interventionsService
+          .updateActionPlanAppointment(res.locals.user.token.accessToken, req.params.id, sessionNumber, paramsForUpdate)
+          .then(),
+      createPresenter: (appointment, formError, userInputData, serverError) =>
+        new EditSessionPresenter(appointment, formError, userInputData, serverError),
+      redirectTo: `/service-provider/referrals/${actionPlan.referralId}/progress`,
+    })
+  }
 
+  private async scheduleAppointment<AppointmentType>(
```

Comment:
This took me a while to figure out what was going on for so early on a monday morning but it's very clever, nice approach.